---
title: "4th_inference_work"
author: "Jingsong Zhou"
date: "2023-09-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(dplyr)
# read the dataset
dataset <- read.table("my_agam_aging_expr_timeseries_Jun23_2023.tsv", header = TRUE)
```

**1. Improvement about Model for Means**

At this level, the model captures the shared characteristics and systematic variation among genes, considering the effect of both gene-specific intercepts, strain-specific intercepts, and the change in expression mean per unit change in age.

Notice that $\mu_{ij}$ is the j-th entry in the mean vector $\boldsymbol{\mu}_i$. The mean expression value $\mu_{ij}$ for gene j at the i-th age and strains is modeled as:

$$
\mu_{ij} = \alpha_{mean,j} + \beta_{age,j} \cdot \text{age}_i + \sum_{k=1}^{N_{\text{strain}}}\alpha_{j,\text{strain_k}} \cdot \boldsymbol{1_{\text{strain_k}}}
$$

Where:

- $\mu_{ij}$ is the mean expression value of the j-th gene at the i-th age.
- $\alpha_{mean,j}$ is the intercept term, representing the overall mean expression level across ages and strains for j-th gene.
- $\beta_{age,j}$ is the coefficient of the 'age' variable, representing the change in mean expression of j-th gene per unit change in age.
- $N_{\text{strain}}$ is the total number of strains.
- $\alpha_{j,\text{strain_k}}$ is the coefficients associated with the k-th strain, which represents the fixed effect of the k-th strain for j-th gene.
- $\boldsymbol{1_{\text{strain_k}}}$ is the indicator variable for the k-th strain.


In the below modeling approach, we are utilizing individual observations from gene expression data to fit models. Unlike the previous approach that relied on estimated means, this method directly leverages the raw data points, avoiding potential information loss from different replicates.

```{r}
library(dplyr)

# Group the mean_expression by geneID
expression_mean_model <- dataset %>%
  group_by(geneID) %>%
  do(mod = lm(transcription ~ age + factor(strainID), data = .))

# Extract coefficients for each gene
expression_mean_results <- expression_mean_model %>%
  mutate(
    alpha_gene = coef(mod)[1],
    beta_age = coef(mod)[2]
  ) %>%
  select(geneID, alpha_gene, beta_age)
```

**Explanation of Why We Separate Subsets and Fit the Model for Each Gene:**

The model contains only gene-specific effects, we should be able to obtain the same result whether we fit the model after subsetting for a specific gene or fit the model using the entire dataset. In this scenario, the model parameters (coefficients) are specific to each gene, and they are estimated independently of each other. Whether we subset the dataset for a specific gene or use the entire dataset, the estimation process for that specific gene's model remains the same. 

Also, extracting and organizing coefficients for individual genes becomes much simpler when we're working with a smaller subset of models. This facilitates further analysis, such as comparing gene-specific effects across genes or identifying significant associations. 



**2. Alternatives for Variance Models**

**Modeling the Variance of Gene Expression ($\delta_{ij}$):**

$\delta_{ij}$ represents the variance of gene expression for the j-th gene at the i-th age, which is the jj-th entry of the diagonal matrix $\text{diag}(\boldsymbol{\Delta}_{i})$. To model the variance of gene expression for each gene over age, we assume a linear relationship with age:

$$
\delta_{ij} = \alpha_{var,j} + \beta_{\text{age_var_j}} \cdot \text{age}_i + \sum_{k=1}^{N_{\text{strain}}}\alpha_{var,j,\text{strain_k}} \cdot \boldsymbol{1_{\text{strain_k}}}
$$

where:

- $\delta_{ij}$ represents the variance of gene expression for the j-th gene at the i-th age.
- $\alpha_{var,j}$ represents the gene-specific intercept for the j-th gene.
- $\beta_{\text{age_var_j}}$ represents the coefficient capturing the change in variance per unit change in age.
- $N_{\text{strain}}$ is the total number of strains.
- $\alpha_{var,j,\text{strain_k}}$ is the coefficients associated with the k-th strain, which represents the fixed effect of the k-th strain for j-th gene.
- $\boldsymbol{1_{\text{strain_k}}}$ is the indicator variable for the k-th strain.

```{r}
# Group by age and geneID and estimate the variance of gene expression
variance_gene <- dataset %>%
  group_by(age, geneID, strainID) %>%
  summarize(variance_expression = var(transcription))
```


```{r}
#'variance_gene' is the dataframe with columns: age, geneID, variance_expression
variance_gene_model <- variance_gene %>%
  group_by(geneID) %>%
  do(mod = lm(variance_expression ~ age + factor(strainID), data = .))

# Extract alpha_j (intercept), beta_age_var_j (slope), and p-value coefficients
variance_model_result <- variance_gene_model %>%
  mutate(alpha_j = coef(mod)[1], beta_age_var_j = coef(mod)[2], p_value = summary(mod)$coefficients[2, 4]) %>%
  select(geneID, alpha_j, beta_age_var_j, p_value)
```

```{r}
# Calculate the mean and standard deviation of beta_age_var_j
mean_slope <- mean(variance_model_result$beta_age_var_j)
sd_slope <- sd(variance_model_result$beta_age_var_j)

# Set a Z-score threshold for outlier detection (e.g., Z > 3 for extreme outliers)
z_threshold <- 3

# Calculate Z-scores for each gene's slope
variance_model_result <- variance_model_result %>%
  mutate(z_score = (beta_age_var_j - mean_slope) / sd_slope)

# Identify geneIDs as outliers based on the Z-score threshold
outliers <- variance_model_result %>%
  filter(abs(z_score) > z_threshold)
```

```{r}
library(ggplot2)
# Create a histogram for the overall distribution
histogram <- ggplot(data = variance_model_result, aes(x = beta_age_var_j)) +
  geom_histogram(binwidth = 0.01, fill = "blue", alpha = 0.7) +
  labs(title = "Histogram of Slopes for Age Overall",
       y = "Frequency") +
  scale_x_continuous(breaks = seq(-0.1, 0.1, by = 0.05),limits = c(-0.12,0.12))

# Create a histogram for the outliers
histogram_outliers <- ggplot(data = outliers, aes(x = beta_age_var_j)) +
  geom_histogram(binwidth = 0.01, fill = "red", alpha = 0.7) +
  labs(title = "Histogram for Outliers",
       y = "Frequency")

# Display both histograms side by side
library(gridExtra)
grid.arrange(histogram, histogram_outliers, ncol = 2)
```


**3. Annotation resource use**

Here are the two links that can lead us to the resource we are interested, especially about annotations of genes.
[Ensemble](https://rapid.ensembl.org/Anopheles_gambiae_GCA_943734735.1/Info/Index?db=core)
[Vectorbase](https://vectorbase.org/vectorbase/app/search?q=Anopheles%20gambiae%20PEST&documentType=gene&organisms=Anopheles%20gambiae%20PEST)

And I also find a link may includes related information:
[NIH](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000005575.2/)


