{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "92880ec8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "00931a50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcription</th>\n",
       "      <th>strainID</th>\n",
       "      <th>age</th>\n",
       "      <th>geneID</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>experimentID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10.617924</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>AGAP005098</td>\n",
       "      <td>AGAP005098-RA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9.614043</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>AGAP005098</td>\n",
       "      <td>AGAP005098-RA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9.930478</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>AGAP005098</td>\n",
       "      <td>AGAP005098-RA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10.160402</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>AGAP005098</td>\n",
       "      <td>AGAP005098-RA</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10.482997</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>AGAP005098</td>\n",
       "      <td>AGAP005098-RA</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   transcription  strainID  age      geneID   transcriptID  experimentID\n",
       "0      10.617924         0   14  AGAP005098  AGAP005098-RA             1\n",
       "1       9.614043         0   14  AGAP005098  AGAP005098-RA             2\n",
       "2       9.930478         0   14  AGAP005098  AGAP005098-RA             3\n",
       "3      10.160402         1   14  AGAP005098  AGAP005098-RA             4\n",
       "4      10.482997         1   14  AGAP005098  AGAP005098-RA             5"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv('my_agam_aging_expr_timeseries_Jun23_2023.tsv',sep='\\t')\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5250e9a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>age</th>\n",
       "      <th>strainID</th>\n",
       "      <th>variance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAP000002</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGAP000002</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>0.055051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGAP000002</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGAP000002</td>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGAP000002</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0.037857</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       geneID  age  strainID  variance\n",
       "0  AGAP000002   14         0  0.000509\n",
       "1  AGAP000002   14         1  0.055051\n",
       "2  AGAP000002   14         2  0.000855\n",
       "3  AGAP000002   14         3  0.000642\n",
       "4  AGAP000002   21         0  0.037857"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# First, we need to compute the variance of gene expression for each gene at each age.\n",
    "# This involves grouping the data by 'geneID' and 'age', then calculating the variance of the 'transcription' values.\n",
    "\n",
    "# Grouping the data by 'geneID' and 'age' and calculating the variance\n",
    "gene_expression_variance = data.groupby(['geneID', 'age', 'strainID'])['transcription'].var()\n",
    "\n",
    "# Resetting the index to make 'geneID', 'age', and 'strainID' columns again\n",
    "gene_expression_variance = gene_expression_variance.reset_index()\n",
    "\n",
    "gene_expression_variance.rename(columns={'transcription': 'variance'}, inplace=True)\n",
    "\n",
    "# Displaying the first few rows of the newly created dataframe\n",
    "gene_expression_variance.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3d96e6f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\86187\\anaconda3\\lib\\site-packages\\statsmodels\\regression\\mixed_linear_model.py:2237: ConvergenceWarning: The MLE may be on the boundary of the parameter space.\n",
      "  warnings.warn(msg, ConvergenceWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          Mixed Linear Model Regression Results\n",
      "=========================================================\n",
      "Model:            MixedLM Dependent Variable: variance   \n",
      "No. Observations: 213282  Method:             REML       \n",
      "No. Groups:       12546   Scale:              0.1185     \n",
      "Min. group size:  17      Log-Likelihood:     -98803.5904\n",
      "Max. group size:  17      Converged:          Yes        \n",
      "Mean group size:  17.0                                   \n",
      "---------------------------------------------------------\n",
      "                Coef. Std.Err.   z    P>|z| [0.025 0.975]\n",
      "---------------------------------------------------------\n",
      "Intercept       0.192    0.004 42.909 0.000  0.183  0.200\n",
      "age             0.001    0.000 10.133 0.000  0.001  0.001\n",
      "Group Var       0.189    0.007                           \n",
      "strainID Var    0.004    0.001                           \n",
      "=========================================================\n",
      "\n"
     ]
    }
   ],
   "source": [
    "model_with_strain = smf.mixedlm(\"variance ~ age\", data=gene_expression_variance, \n",
    "                                groups=gene_expression_variance[\"geneID\"], \n",
    "                                re_formula=\"~1\", \n",
    "                                vc_formula={\"strainID\": \"0 + C(strainID)\"})\n",
    "result_with_strain = model_with_strain.fit()\n",
    "\n",
    "# Print the summary of the model with strain effect\n",
    "print(result_with_strain.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7c01aceb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Coef.</th>\n",
       "      <th>Std.Err.</th>\n",
       "      <th>z</th>\n",
       "      <th>P&gt;|z|</th>\n",
       "      <th>[0.025</th>\n",
       "      <th>0.975]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Intercept</th>\n",
       "      <td>0.192</td>\n",
       "      <td>0.004</td>\n",
       "      <td>42.909</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.183</td>\n",
       "      <td>0.200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>age</th>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>10.133</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Group Var</th>\n",
       "      <td>0.189</td>\n",
       "      <td>0.007</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>strainID Var</th>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Coef. Std.Err.       z  P>|z| [0.025 0.975]\n",
       "Intercept     0.192    0.004  42.909  0.000  0.183  0.200\n",
       "age           0.001    0.000  10.133  0.000  0.001  0.001\n",
       "Group Var     0.189    0.007                             \n",
       "strainID Var  0.004    0.001                             "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_with_strain.summary().tables[1]\n",
    "# Group Var (Random Effect for GeneID)\n",
    "# Coef (0.189): This is the estimated variance of the random intercepts for genes, \n",
    "# indicating how much gene expression variance varies from one gene to another.\n",
    "\n",
    "# StrainID Var (Random Effect for StrainID)\n",
    "# Similarly as Group Var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "18824164",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group</th>\n",
       "      <th>strainID[C(strainID)[0]]</th>\n",
       "      <th>strainID[C(strainID)[1]]</th>\n",
       "      <th>strainID[C(strainID)[2]]</th>\n",
       "      <th>strainID[C(strainID)[3]]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AGAP000002</th>\n",
       "      <td>-0.184895</td>\n",
       "      <td>-0.002178</td>\n",
       "      <td>-0.000039</td>\n",
       "      <td>-0.000695</td>\n",
       "      <td>-0.001402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAP000005</th>\n",
       "      <td>-0.117533</td>\n",
       "      <td>-0.008847</td>\n",
       "      <td>-0.005187</td>\n",
       "      <td>0.016571</td>\n",
       "      <td>-0.005279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAP000007</th>\n",
       "      <td>0.013649</td>\n",
       "      <td>0.005128</td>\n",
       "      <td>-0.009891</td>\n",
       "      <td>0.012181</td>\n",
       "      <td>-0.007100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAP000008</th>\n",
       "      <td>-0.100636</td>\n",
       "      <td>-0.009514</td>\n",
       "      <td>-0.009032</td>\n",
       "      <td>0.001944</td>\n",
       "      <td>0.014254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAP000009</th>\n",
       "      <td>-0.100822</td>\n",
       "      <td>-0.015004</td>\n",
       "      <td>-0.007024</td>\n",
       "      <td>0.010959</td>\n",
       "      <td>0.008715</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Group  strainID[C(strainID)[0]]  strainID[C(strainID)[1]]  \\\n",
       "AGAP000002 -0.184895                 -0.002178                 -0.000039   \n",
       "AGAP000005 -0.117533                 -0.008847                 -0.005187   \n",
       "AGAP000007  0.013649                  0.005128                 -0.009891   \n",
       "AGAP000008 -0.100636                 -0.009514                 -0.009032   \n",
       "AGAP000009 -0.100822                 -0.015004                 -0.007024   \n",
       "\n",
       "            strainID[C(strainID)[2]]  strainID[C(strainID)[3]]  \n",
       "AGAP000002                 -0.000695                 -0.001402  \n",
       "AGAP000005                  0.016571                 -0.005279  \n",
       "AGAP000007                  0.012181                 -0.007100  \n",
       "AGAP000008                  0.001944                  0.014254  \n",
       "AGAP000009                  0.010959                  0.008715  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extracting random effects\n",
    "random_effects = result_with_strain.random_effects\n",
    "# Converting the dictionary of random effects to a DataFrame\n",
    "random_effects_df = pd.DataFrame.from_dict(random_effects, orient='index')\n",
    "random_effects_df.head()\n",
    "\n",
    "#The column \"Group\" in random_effects_df represents these slopes over age for each group(gene in our context)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94890492",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2900cf37",
   "metadata": {},
   "source": [
    "I also conducted a Gene Ontology analysis using a tool named DAVID.Below are the results for genes with significantly increasing variance over age.\n",
    "\n",
    "Here's a brief overview of the key columns in the DAVID summary chart:\n",
    "\n",
    "1. **Category**: This indicates the type of Gene Ontology (GO) term or other categorization, such as GOTERM_BP_DIRECT (Biological Process), GOTERM_MF_DIRECT (Molecular Function), or GOTERM_CC_DIRECT (Cellular Component).\n",
    "\n",
    "2. **Term**: The specific GO term or pathway identified in the analysis. It usually contains the GO ID and a brief description (e.g., \"GO:0005549~odorant binding\").\n",
    "\n",
    "3. **Count**: The number of genes from the list that are associated with this particular term.\n",
    "\n",
    "4. **%**: The percentage of genes from the list that fall into this category.\n",
    "\n",
    "5. **PValue**: The p-value indicating the statistical significance of the association between the gene list and the GO term. Lower values indicate a more significant association.\n",
    "\n",
    "6. **Genes**: The specific genes from the list that are associated with this GO term.\n",
    "\n",
    "7. **List Total**: The total number of genes in the list.\n",
    "\n",
    "8. **Pop Hits**: The number of genes in the background population (like the entire genome) that are associated with this term.\n",
    "\n",
    "9. **Pop Total**: The total number of genes in the background population.\n",
    "\n",
    "10. **Fold Enrichment**: Indicates how much more common the GO term is in your list compared to the background population. Higher values suggest a stronger association.\n",
    "\n",
    "11. **Bonferroni, Benjamini, FDR**: These are different methods of adjusting the p-value for multiple testing. They help control the false discovery rate when many hypotheses are tested simultaneously.\n",
    "\n",
    "- Look for trends or common themes among the highly significant GO terms. Do they cluster around certain biological processes or functions?\n",
    "- Identify the most enriched pathways or processes. These are the ones where the genes are overrepresented compared to what would be expected by chance.\n",
    "- Since we're studying aging, we are particularly interested in GO terms related to cell senescence, DNA repair, or metabolic processes.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6968ecae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Category</th>\n",
       "      <th>Term</th>\n",
       "      <th>Count</th>\n",
       "      <th>%</th>\n",
       "      <th>PValue</th>\n",
       "      <th>Genes</th>\n",
       "      <th>List Total</th>\n",
       "      <th>Pop Hits</th>\n",
       "      <th>Pop Total</th>\n",
       "      <th>Fold Enrichment</th>\n",
       "      <th>Bonferroni</th>\n",
       "      <th>Benjamini</th>\n",
       "      <th>FDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOTERM_BP_DIRECT</td>\n",
       "      <td>GO:0050911~detection of chemical stimulus invo...</td>\n",
       "      <td>67</td>\n",
       "      <td>1.686383</td>\n",
       "      <td>8.183479e-31</td>\n",
       "      <td>AGAP009520, AGAP012854, AGAP009640, AGAP004355...</td>\n",
       "      <td>1554</td>\n",
       "      <td>80</td>\n",
       "      <td>7042</td>\n",
       "      <td>3.795158</td>\n",
       "      <td>8.478084e-28</td>\n",
       "      <td>8.469901e-28</td>\n",
       "      <td>8.420800e-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOTERM_MF_DIRECT</td>\n",
       "      <td>GO:0005549~odorant binding</td>\n",
       "      <td>101</td>\n",
       "      <td>2.542160</td>\n",
       "      <td>4.716560e-29</td>\n",
       "      <td>AGAP009520, AGAP009640, AGAP012659, AGAP012658...</td>\n",
       "      <td>1885</td>\n",
       "      <td>149</td>\n",
       "      <td>7787</td>\n",
       "      <td>2.800231</td>\n",
       "      <td>3.358191e-26</td>\n",
       "      <td>3.358191e-26</td>\n",
       "      <td>3.292159e-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOTERM_MF_DIRECT</td>\n",
       "      <td>GO:0004984~olfactory receptor activity</td>\n",
       "      <td>68</td>\n",
       "      <td>1.711553</td>\n",
       "      <td>3.673612e-28</td>\n",
       "      <td>AGAP009520, AGAP012854, AGAP009640, AGAP004355...</td>\n",
       "      <td>1885</td>\n",
       "      <td>82</td>\n",
       "      <td>7787</td>\n",
       "      <td>3.425736</td>\n",
       "      <td>2.615612e-25</td>\n",
       "      <td>1.307806e-25</td>\n",
       "      <td>1.282091e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOTERM_BP_DIRECT</td>\n",
       "      <td>GO:0050909~sensory perception of taste</td>\n",
       "      <td>46</td>\n",
       "      <td>1.157815</td>\n",
       "      <td>2.196892e-22</td>\n",
       "      <td>AGAP010195, AGAP009803, AGAP006875, AGAP001120...</td>\n",
       "      <td>1554</td>\n",
       "      <td>53</td>\n",
       "      <td>7042</td>\n",
       "      <td>3.933027</td>\n",
       "      <td>2.275980e-19</td>\n",
       "      <td>1.136892e-19</td>\n",
       "      <td>1.130301e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOTERM_CC_DIRECT</td>\n",
       "      <td>GO:0005886~plasma membrane</td>\n",
       "      <td>276</td>\n",
       "      <td>6.946892</td>\n",
       "      <td>1.321277e-19</td>\n",
       "      <td>AGAP012891, AGAP012656, AGAP004355, AGAP009803...</td>\n",
       "      <td>1945</td>\n",
       "      <td>731</td>\n",
       "      <td>8256</td>\n",
       "      <td>1.602661</td>\n",
       "      <td>5.245468e-17</td>\n",
       "      <td>5.245468e-17</td>\n",
       "      <td>5.205830e-17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Category                                               Term  Count  \\\n",
       "0  GOTERM_BP_DIRECT  GO:0050911~detection of chemical stimulus invo...     67   \n",
       "1  GOTERM_MF_DIRECT                         GO:0005549~odorant binding    101   \n",
       "2  GOTERM_MF_DIRECT             GO:0004984~olfactory receptor activity     68   \n",
       "3  GOTERM_BP_DIRECT             GO:0050909~sensory perception of taste     46   \n",
       "4  GOTERM_CC_DIRECT                         GO:0005886~plasma membrane    276   \n",
       "\n",
       "          %        PValue                                              Genes  \\\n",
       "0  1.686383  8.183479e-31  AGAP009520, AGAP012854, AGAP009640, AGAP004355...   \n",
       "1  2.542160  4.716560e-29  AGAP009520, AGAP009640, AGAP012659, AGAP012658...   \n",
       "2  1.711553  3.673612e-28  AGAP009520, AGAP012854, AGAP009640, AGAP004355...   \n",
       "3  1.157815  2.196892e-22  AGAP010195, AGAP009803, AGAP006875, AGAP001120...   \n",
       "4  6.946892  1.321277e-19  AGAP012891, AGAP012656, AGAP004355, AGAP009803...   \n",
       "\n",
       "   List Total  Pop Hits  Pop Total  Fold Enrichment    Bonferroni  \\\n",
       "0        1554        80       7042         3.795158  8.478084e-28   \n",
       "1        1885       149       7787         2.800231  3.358191e-26   \n",
       "2        1885        82       7787         3.425736  2.615612e-25   \n",
       "3        1554        53       7042         3.933027  2.275980e-19   \n",
       "4        1945       731       8256         1.602661  5.245468e-17   \n",
       "\n",
       "      Benjamini           FDR  \n",
       "0  8.469901e-28  8.420800e-28  \n",
       "1  3.358191e-26  3.292159e-26  \n",
       "2  1.307806e-25  1.282091e-25  \n",
       "3  1.136892e-19  1.130301e-19  \n",
       "4  5.245468e-17  5.205830e-17  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = 'chart_Go.txt'\n",
    "go_data = pd.read_csv(file_path, sep=\"\\t\")\n",
    "\n",
    "go_data.head() \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5da4cc89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Category</th>\n",
       "      <th>Term</th>\n",
       "      <th>Count</th>\n",
       "      <th>%</th>\n",
       "      <th>PValue</th>\n",
       "      <th>Genes</th>\n",
       "      <th>List Total</th>\n",
       "      <th>Pop Hits</th>\n",
       "      <th>Pop Total</th>\n",
       "      <th>Fold Enrichment</th>\n",
       "      <th>Bonferroni</th>\n",
       "      <th>Benjamini</th>\n",
       "      <th>FDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>GOTERM_BP_DIRECT</td>\n",
       "      <td>GO:0046177~D-gluconate catabolic process</td>\n",
       "      <td>6</td>\n",
       "      <td>0.151019</td>\n",
       "      <td>2.543313e-03</td>\n",
       "      <td>AGAP010329, AGAP004687, AGAP012497, AGAP004197...</td>\n",
       "      <td>1554</td>\n",
       "      <td>6</td>\n",
       "      <td>7042</td>\n",
       "      <td>4.531532</td>\n",
       "      <td>9.285122e-01</td>\n",
       "      <td>2.632328e-01</td>\n",
       "      <td>2.617069e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GOTERM_MF_DIRECT</td>\n",
       "      <td>GO:0004616~phosphogluconate dehydrogenase (dec...</td>\n",
       "      <td>12</td>\n",
       "      <td>0.302039</td>\n",
       "      <td>1.520606e-06</td>\n",
       "      <td>AGAP010329, AGAP004687, AGAP004676, AGAP012497...</td>\n",
       "      <td>1885</td>\n",
       "      <td>12</td>\n",
       "      <td>7787</td>\n",
       "      <td>4.131034</td>\n",
       "      <td>1.082086e-03</td>\n",
       "      <td>2.706678e-04</td>\n",
       "      <td>2.653457e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOTERM_BP_DIRECT</td>\n",
       "      <td>GO:0050909~sensory perception of taste</td>\n",
       "      <td>46</td>\n",
       "      <td>1.157815</td>\n",
       "      <td>2.196892e-22</td>\n",
       "      <td>AGAP010195, AGAP009803, AGAP006875, AGAP001120...</td>\n",
       "      <td>1554</td>\n",
       "      <td>53</td>\n",
       "      <td>7042</td>\n",
       "      <td>3.933027</td>\n",
       "      <td>2.275980e-19</td>\n",
       "      <td>1.136892e-19</td>\n",
       "      <td>1.130301e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>GOTERM_BP_DIRECT</td>\n",
       "      <td>GO:0009051~pentose-phosphate shunt, oxidative ...</td>\n",
       "      <td>6</td>\n",
       "      <td>0.151019</td>\n",
       "      <td>7.280058e-03</td>\n",
       "      <td>AGAP010329, AGAP004687, AGAP012497, AGAP004197...</td>\n",
       "      <td>1554</td>\n",
       "      <td>7</td>\n",
       "      <td>7042</td>\n",
       "      <td>3.884170</td>\n",
       "      <td>9.994842e-01</td>\n",
       "      <td>4.709287e-01</td>\n",
       "      <td>4.681987e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOTERM_BP_DIRECT</td>\n",
       "      <td>GO:0050911~detection of chemical stimulus invo...</td>\n",
       "      <td>67</td>\n",
       "      <td>1.686383</td>\n",
       "      <td>8.183479e-31</td>\n",
       "      <td>AGAP009520, AGAP012854, AGAP009640, AGAP004355...</td>\n",
       "      <td>1554</td>\n",
       "      <td>80</td>\n",
       "      <td>7042</td>\n",
       "      <td>3.795158</td>\n",
       "      <td>8.478084e-28</td>\n",
       "      <td>8.469901e-28</td>\n",
       "      <td>8.420800e-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>GOTERM_BP_DIRECT</td>\n",
       "      <td>GO:0006098~pentose-phosphate shunt</td>\n",
       "      <td>8</td>\n",
       "      <td>0.201359</td>\n",
       "      <td>1.591628e-03</td>\n",
       "      <td>AGAP010329, AGAP004676, AGAP012542, AGAP012583...</td>\n",
       "      <td>1554</td>\n",
       "      <td>10</td>\n",
       "      <td>7042</td>\n",
       "      <td>3.625225</td>\n",
       "      <td>8.079962e-01</td>\n",
       "      <td>1.830372e-01</td>\n",
       "      <td>1.819761e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>GOTERM_BP_DIRECT</td>\n",
       "      <td>GO:0050916~sensory perception of sweet taste</td>\n",
       "      <td>7</td>\n",
       "      <td>0.176189</td>\n",
       "      <td>5.143217e-03</td>\n",
       "      <td>AGAP003255, AGAP003256, AGAP003258, AGAP003253...</td>\n",
       "      <td>1554</td>\n",
       "      <td>9</td>\n",
       "      <td>7042</td>\n",
       "      <td>3.524525</td>\n",
       "      <td>9.952143e-01</td>\n",
       "      <td>3.919050e-01</td>\n",
       "      <td>3.896331e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOTERM_MF_DIRECT</td>\n",
       "      <td>GO:0004984~olfactory receptor activity</td>\n",
       "      <td>68</td>\n",
       "      <td>1.711553</td>\n",
       "      <td>3.673612e-28</td>\n",
       "      <td>AGAP009520, AGAP012854, AGAP009640, AGAP004355...</td>\n",
       "      <td>1885</td>\n",
       "      <td>82</td>\n",
       "      <td>7787</td>\n",
       "      <td>3.425736</td>\n",
       "      <td>2.615612e-25</td>\n",
       "      <td>1.307806e-25</td>\n",
       "      <td>1.282091e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>GOTERM_BP_DIRECT</td>\n",
       "      <td>GO:0060285~cilium-dependent cell motility</td>\n",
       "      <td>6</td>\n",
       "      <td>0.151019</td>\n",
       "      <td>1.590046e-02</td>\n",
       "      <td>AGAP002143, AGAP001293, AGAP006596, AGAP003805...</td>\n",
       "      <td>1554</td>\n",
       "      <td>8</td>\n",
       "      <td>7042</td>\n",
       "      <td>3.398649</td>\n",
       "      <td>9.999999e-01</td>\n",
       "      <td>7.836657e-01</td>\n",
       "      <td>7.791227e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>GOTERM_MF_DIRECT</td>\n",
       "      <td>GO:0004620~phospholipase activity</td>\n",
       "      <td>8</td>\n",
       "      <td>0.201359</td>\n",
       "      <td>2.844505e-03</td>\n",
       "      <td>AGAP009956, AGAP005971, AGAP006396, AGAP011512...</td>\n",
       "      <td>1885</td>\n",
       "      <td>10</td>\n",
       "      <td>7787</td>\n",
       "      <td>3.304828</td>\n",
       "      <td>8.684244e-01</td>\n",
       "      <td>1.125160e-01</td>\n",
       "      <td>1.103036e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Category                                               Term  \\\n",
       "41  GOTERM_BP_DIRECT           GO:0046177~D-gluconate catabolic process   \n",
       "14  GOTERM_MF_DIRECT  GO:0004616~phosphogluconate dehydrogenase (dec...   \n",
       "3   GOTERM_BP_DIRECT             GO:0050909~sensory perception of taste   \n",
       "53  GOTERM_BP_DIRECT  GO:0009051~pentose-phosphate shunt, oxidative ...   \n",
       "0   GOTERM_BP_DIRECT  GO:0050911~detection of chemical stimulus invo...   \n",
       "37  GOTERM_BP_DIRECT                 GO:0006098~pentose-phosphate shunt   \n",
       "49  GOTERM_BP_DIRECT       GO:0050916~sensory perception of sweet taste   \n",
       "2   GOTERM_MF_DIRECT             GO:0004984~olfactory receptor activity   \n",
       "70  GOTERM_BP_DIRECT          GO:0060285~cilium-dependent cell motility   \n",
       "42  GOTERM_MF_DIRECT                  GO:0004620~phospholipase activity   \n",
       "\n",
       "    Count         %        PValue  \\\n",
       "41      6  0.151019  2.543313e-03   \n",
       "14     12  0.302039  1.520606e-06   \n",
       "3      46  1.157815  2.196892e-22   \n",
       "53      6  0.151019  7.280058e-03   \n",
       "0      67  1.686383  8.183479e-31   \n",
       "37      8  0.201359  1.591628e-03   \n",
       "49      7  0.176189  5.143217e-03   \n",
       "2      68  1.711553  3.673612e-28   \n",
       "70      6  0.151019  1.590046e-02   \n",
       "42      8  0.201359  2.844505e-03   \n",
       "\n",
       "                                                Genes  List Total  Pop Hits  \\\n",
       "41  AGAP010329, AGAP004687, AGAP012497, AGAP004197...        1554         6   \n",
       "14  AGAP010329, AGAP004687, AGAP004676, AGAP012497...        1885        12   \n",
       "3   AGAP010195, AGAP009803, AGAP006875, AGAP001120...        1554        53   \n",
       "53  AGAP010329, AGAP004687, AGAP012497, AGAP004197...        1554         7   \n",
       "0   AGAP009520, AGAP012854, AGAP009640, AGAP004355...        1554        80   \n",
       "37  AGAP010329, AGAP004676, AGAP012542, AGAP012583...        1554        10   \n",
       "49  AGAP003255, AGAP003256, AGAP003258, AGAP003253...        1554         9   \n",
       "2   AGAP009520, AGAP012854, AGAP009640, AGAP004355...        1885        82   \n",
       "70  AGAP002143, AGAP001293, AGAP006596, AGAP003805...        1554         8   \n",
       "42  AGAP009956, AGAP005971, AGAP006396, AGAP011512...        1885        10   \n",
       "\n",
       "    Pop Total  Fold Enrichment    Bonferroni     Benjamini           FDR  \n",
       "41       7042         4.531532  9.285122e-01  2.632328e-01  2.617069e-01  \n",
       "14       7787         4.131034  1.082086e-03  2.706678e-04  2.653457e-04  \n",
       "3        7042         3.933027  2.275980e-19  1.136892e-19  1.130301e-19  \n",
       "53       7042         3.884170  9.994842e-01  4.709287e-01  4.681987e-01  \n",
       "0        7042         3.795158  8.478084e-28  8.469901e-28  8.420800e-28  \n",
       "37       7042         3.625225  8.079962e-01  1.830372e-01  1.819761e-01  \n",
       "49       7042         3.524525  9.952143e-01  3.919050e-01  3.896331e-01  \n",
       "2        7787         3.425736  2.615612e-25  1.307806e-25  1.282091e-25  \n",
       "70       7042         3.398649  9.999999e-01  7.836657e-01  7.791227e-01  \n",
       "42       7787         3.304828  8.684244e-01  1.125160e-01  1.103036e-01  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Analysis of the DAVID summary data\n",
    "\n",
    "# Filter the data to focus on highly significant terms\n",
    "# We'll use a p-value threshold of 0.05 and fold enrichment greater than 1.5 as an example\n",
    "significant_terms = go_data[(go_data['PValue'] < 0.05) & (go_data['Fold Enrichment'] > 1.5)]\n",
    "\n",
    "# Analysis 1: Identifying the most enriched GO terms\n",
    "most_enriched_terms = significant_terms.sort_values(by='Fold Enrichment', ascending=False).head(10)\n",
    "\n",
    "most_enriched_terms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11a4df99",
   "metadata": {},
   "source": [
    "These terms suggest that processes related to sensory perception (taste and smell), metabolism (particularly involving pentose-phosphate shunt and glu conate catabolism), and certain molecular functions like olfactory receptor activity and phospholipase activity are highly represented among genes with increased variance in expression with age."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ccbe117a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AGAP010790    11\n",
       "AGAP005637     9\n",
       "AGAP005636     9\n",
       "AGAP005638     9\n",
       "AGAP009999     9\n",
       "AGAP006221     9\n",
       "AGAP006224     9\n",
       "AGAP011859     9\n",
       "AGAP009137     8\n",
       "AGAP009519     8\n",
       "dtype: int64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Analysis 2\n",
    "# Extracting and analyzing the genes from the DAVID summary data to identify frequently occurring genes\n",
    "\n",
    "# Extracting the genes from each GO term\n",
    "genes_list = go_data['Genes'].str.split(',').tolist()\n",
    "# Flattening the list\n",
    "all_genes = [gene.strip() for sublist in genes_list for gene in sublist]\n",
    "\n",
    "# Counting the frequency of each gene\n",
    "gene_frequency = pd.Series(all_genes).value_counts()\n",
    "\n",
    "# Display the most frequently occurring genes\n",
    "most_frequent_genes = gene_frequency.head(10)  # Top 10 for example\n",
    "most_frequent_genes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a001db39",
   "metadata": {},
   "source": [
    "These genes have frequent association with the enriched GO terms in the DAVID summary. They might play significant roles in the biological processes or pathways impacted by aging, as indicated by their repeated occurrence across different GO categories."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
